intermediate_source/process_group_cpp_extension_tutorial.rst

```diff
@@ -0,0 +1,284 @@
+``ProcessGroupNCCL``, ``ProcessGroupGloo``, and ``ProcessGroupMPI``. However,
+beyond these three backends, there are also other communication libraries
+(e.g., `UCC <https://github.com/openucx/ucc>`__,
+`OneCCL <https://github.com/oneapi-src/oneCCL>`__), different types of hardware
+(e.g., `TPU <https://cloud.google.com/tpu>`__,
+`Trainum <https://aws.amazon.com/machine-learning/trainium/>`__), and emerging
+communication algorithms (e.g.,
+`Herring <https://www.amazon.science/publications/herring-rethinking-the-parameter-server-at-scale-for-the-cloud>`__,
+`Reduction Server <https://cloud.google.com/blog/topics/developers-practitioners/optimize-training-performance-reduction-server-vertex-ai>`__).
+Therefore, the distributed package exposed extension APIs to allow customizing
+collective communication backends.
+
+
+The 4 steps below show how to implement a dummy ``ProcessGroup`` backend
+and use that in Python application code. Please note that this tutorial focuses
+on demonstrating the extension APIs, instead of developing a functioning
+communication backend. Hence, the ``dummy`` backend just covers a subset of the
+APIs (``all_reduce`` and ``all_gather``), and simply sets the values of tensors
+to 0.
+
+
+Step 1: Implement a Subclass of ``ProcessGroup``
+------------------------------------------------
+
+This first step is to implement a ``ProcessGroup`` subclass that overrides
+target collective communication APIs and runs the custom communication algorithm.
+The extension also needs to implement a ``ProcessGroup::Work`` subclass, which
+serves as a future of communication results and allows asynchronous execution in
+application code. If the extension uses third-party libraries, it can
+include the headers and call into the library APIs from the ``ProcessGroupDummy``
+subclass. The two code snippets below present the implementation of ``dummy.h`` and
+``dummy.cpp``. See the `dummy collectives <https://github.com/mrshenli/dummy_collectives>`__
+repository for the full implementation.
+
+.. code-block:: cpp
+
+    // file name: dummy.hpp
+    #include <torch/python.h>
+
+    #include <c10d/ProcessGroup.hpp>
+    #include <c10d/Store.hpp>
+    #include <c10d/Types.hpp>
+    #include <c10d/Utils.hpp>
+
+    #include <pybind11/chrono.h>
+
+    namespace c10d {
+
+    class ProcessGroupDummy : public ProcessGroup {
+      public:
+
+        class WorkDummy : public ProcessGroup::Work {
+          public:
+            WorkDummy(
+                OpType opType,
+                c10::intrusive_ptr<c10::ivalue::Future> future) // future of the output
+                : ProcessGroup::Work(
+                    -1, // rank, only used by recvAnySource, irrelevant in this demo
+                    opType),
+                future_(std::move(future)) {}
+            bool isCompleted() override;
+            bool isSuccess() const override;
+            bool wait(std::chrono::milliseconds timeout = kUnsetTimeout) override;
```

Comment:
Is it fine for tutorial purposes for these to not be implemented anywhere in the tutorial?

Reply:
good point, let me remove those and add a comment to mention full implementation is in the repo